File: javascript/monothread-sort/quick-sort.js
function sort(array) {
  quickSort(array, 0, array.length - 1);
}

function quickSort(array, low, high) {
  if (low < high) {
    const pi = partition(array, low, high);

    quickSort(array, low, pi - 1);
    quickSort(array, pi + 1, high);
  }
}

function partition(array, low, high) {
  const pivot = array[high];

  let i = low - 1;

  for (let j = low; j <= high - 1; j++) {
    if (array[j] < pivot) {
      i++;
      const aux = array[i];
      array[i] = array[j];
      array[j] = aux;
    }
  }

  const aux = array[i + 1];
  array[i + 1] = array[high];
  array[high] = aux;

  return i + 1;
}

module.exports = {
  sort,
};

File: python/multithread/quick_sort.py
from constants import THREADS_NUMBER
import threading


class QuickSort:
    def sort(self, array: list):
        self.threads_count = 0

        self.quickSort(array, 0, len(array)-1)

    def quickSort(self, array: list, low: int, high: int):
        if low < high:
            pi = self.partition(array, low, high)

            threads = []

            if self.threads_count < THREADS_NUMBER:
                self.threads_count += 1

                thread = threading.Thread(target=self.quickSort,
                                          args=(array, low, pi-1))
                threads.append(thread)
                thread.start()
            else:
                self.quickSort(array, low, pi-1)

            if self.threads_count < THREADS_NUMBER:
                self.threads_count += 1

                thread = threading.Thread(target=self.quickSort,
                                          args=(array, pi+1, high))
                threads.append(thread)
                thread.start()
            else:
                self.quickSort(array, pi+1, high)

            for thread in threads:
                thread.join()

            self.threads_count -= len(threads)

    def partition(self, array: list, low: int, high: int):
        pivot = array[high]

        i = low - 1
        j = low
        while j <= high-1:
            if array[j] < pivot:
                i += 1
                array[i], array[j] = array[j], array[i]

            j += 1

        array[i+1], array[high] = array[high], array[i+1]

        return i + 1
